cat-launcher/src-tauri/src/settings/apply_settings.rs

```diff
@@ -0,0 +1,188 @@
+  get_or_create_user_game_data_dir, get_settings_path,
+  GetUserGameDataDirError,
+};
+use crate::infra::utils::OSNotSupportedError;
+use crate::settings::types::Settings;
+use crate::settings::{FontsConfig, ThemeColors};
+use crate::variants::GameVariant;
+
+#[derive(thiserror::Error, Debug, IntoStaticStr)]
+pub enum ApplySettingsError {
+  #[error("failed to get OS: {0}")]
+  OSNotSupported(#[from] OSNotSupportedError),
+
+  #[error("failed to get user game data dir: {0}")]
+  UserGameDataDirectory(#[from] GetUserGameDataDirError),
+
+  #[error("IO error: {0}")]
+  Io(#[from] io::Error),
+
+  #[error("failed to parse JSON: {0}")]
+  ParseJson(#[from] serde_json::Error),
+
+  #[error("max_backups must be between 0 and 20")]
+  InvalidMaxBackups,
+
+  #[error("parallel_requests must be between 1 and 16")]
+  InvalidParallelRequests,
+
+  #[error("invalid settings path")]
+  InvalidSettingsPath,
+}
+
+pub async fn apply_settings(
+  max_backups: usize,
+  parallel_requests: u16,
+  font_location: Option<String>,
+  theme_colors: Option<ThemeColors>,
+  data_dir: PathBuf,
+  resource_dir: PathBuf,
+) -> Result<(), ApplySettingsError> {
+  if max_backups > 20 {
+    return Err(ApplySettingsError::InvalidMaxBackups);
+  }
+  if !(1..=16).contains(&parallel_requests) {
+    return Err(ApplySettingsError::InvalidParallelRequests);
+  }
+
+  update_settings_file(
+    max_backups,
+    parallel_requests,
+    font_location,
+    theme_colors,
+    &data_dir,
+    &resource_dir,
+  )
+  .await
+}
+
+async fn update_settings_file(
+  max_backups: usize,
+  parallel_requests: u16,
+  font_location: Option<String>,
+  theme_colors: Option<ThemeColors>,
+  data_dir: &Path,
+  resource_dir: &Path,
+) -> Result<(), ApplySettingsError> {
+  let settings_path = get_settings_path(resource_dir);
+  let mut settings: Settings = if settings_path.exists() {
+    let content = fs::read_to_string(&settings_path).await?;
+    serde_json::from_str(&content).unwrap_or_default()
+  } else {
+    Settings::default()
+  };
+
+  settings.max_backups = NonZeroUsize::new(max_backups.max(1))
+    .ok_or(ApplySettingsError::InvalidMaxBackups)?;
+  settings.parallel_requests = NonZeroU16::new(parallel_requests)
+    .ok_or(ApplySettingsError::InvalidParallelRequests)?;
+
+  let settings_json = serde_json::to_string_pretty(&settings)?;
+  fs::write(&settings_path, &settings_json).await?;
+
+  let mut errors = Vec::new();
+
+  if let Some(loc) = font_location {
+    match install_font(&loc, data_dir).await {
+      Ok(_) => settings.font = Some(loc),
+      Err(e) => errors.push(e),
+    }
+  }
+
+  if let Some(colors) = theme_colors {
+    let colors_clone = colors.clone();
+    match install_theme(colors, data_dir).await {
+      Ok(_) => settings.theme = Some(colors_clone),
+      Err(e) => errors.push(e),
+    }
+  }
+
+  let settings_json = serde_json::to_string_pretty(&settings)?;
+  fs::write(&settings_path, settings_json).await?;
+
+  if let Some(e) = errors.pop() {
```

Comment:
In cat-launcher/src-tauri/src/settings/apply_settings.rs around lines 93–115,
the current logic collects multiple errors from install_font and install_theme
but only returns the last one via errors.pop(), losing other failures; change
this to preserve all failures by either returning a combined/aggregate error
variant (create an AggregateInstallError that wraps Vec\<Error> and return that
when errors is non-empty) or, if you prefer simpler behavior, log all subsequent
errors and return the first error immediately (or return the first and include
the rest as context in the returned error); update the error handling so no
failure is silently discarded and adjust call sites/types accordingly.